Don't access the Internet in the check_updates tests

alexwlchan opened the pull request:
Closes https://github.com/rust-lang/rustup/issues/2663

---

Previously, three of the tests in `cli-exact.rs` would try to connect to the Internet, or fail if the Internet wasn't available:

*   `check_updates_with_update`
*   `check_updates_none`
*   `check_updates_some`

This is because all three tests would run `rustup check`, and because the tests weren't overriding the value of `RUSTUP_UPDATE_ROOT`, they'd go out to the Internet to try to download self-updates.

The new helper mimics the behaviour of the `self_update_setup()` helper in the same file.  It creates a mock self-update server in a local directory, and points rustup to the directory using a `file://` URL.  When it tries to get self-updates as part of `rustup check`, it only fetches from that directory.

---

I've tested this patch both with and without an Internet connection, and all the tests pass in both cases.

I've run the Clippy linter, but the only warnings are from code I haven't changed in this patch. Do I need to do anything about that?

<details>
<summary>Output of <code>cargo +nightly clippy --all --all-targets -- -D warnings</code></summary>

```
$ cargo +nightly clippy --all --all-targets -- -D warnings
    Checking rustup v1.23.1 (/Users/alexwlchan/repos/rustup)
error: name `OverrideDB` contains a capitalized acronym
  --> src/config.rs:62:5
   |
62 |     OverrideDB(PathBuf),
   |     ^^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter: `OverrideDb`
   |
   = note: `-D clippy::upper-case-acronyms` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: name `OSProcess` contains a capitalized acronym
   --> src/currentprocess.rs:182:12
    |
182 | pub struct OSProcess {}
    |            ^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter (notice the capitalization): `OsProcess`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: name `OverrideDB` contains a capitalized acronym
  --> src/config.rs:62:5
   |
62 |     OverrideDB(PathBuf),
   |     ^^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter: `OverrideDb`
   |
   = note: `-D clippy::upper-case-acronyms` implied by `-D warnings`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: name `OSProcess` contains a capitalized acronym
   --> src/currentprocess.rs:182:12
    |
182 | pub struct OSProcess {}
    |            ^^^^^^^^^ help: consider making the acronym lowercase, except the initial letter (notice the capitalization): `OsProcess`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: name `IO` contains a capitalized acronym
 --> src/utils/units.rs:6:5
  |
6 |     IO,
  |     ^^ help: consider making the acronym lowercase, except the initial letter (notice the capitalization): `Io`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: name `IO` contains a capitalized acronym
  --> src/utils/units.rs:17:5
   |
17 |     IO(usize, UnitMode),
   |     ^^ help: consider making the acronym lowercase, except the initial letter (notice the capitalization): `Io`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: name `IO` contains a capitalized acronym
 --> src/utils/units.rs:6:5
  |
6 |     IO,
  |     ^^ help: consider making the acronym lowercase, except the initial letter (notice the capitalization): `Io`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: name `IO` contains a capitalized acronym
  --> src/utils/units.rs:17:5
   |
17 |     IO(usize, UnitMode),
   |     ^^ help: consider making the acronym lowercase, except the initial letter (notice the capitalization): `Io`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#upper_case_acronyms

error: aborting due to 4 previous errors

error: could not compile `rustup`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: aborting due to 4 previous errors

error: build failed
```
</details>

h/t @Manishearth for [his tweet linking to Rustup issues](https://twitter.com/ManishEarth/status/1361785243070459904), and @brson for a comment written five years ago (https://github.com/rust-lang/rustup/commit/922babebb3287039ed85aad24153c08e630cca9c / `cli-self-update.rs` / _"Modify the exe so it hashes different"_) that helped me understand some confusing behaviour while I was debugging my changes.

alexwlchan committed:
Don't access the Internet in the check_updates tests

Previously, three of the tests in `cli-exact.rs` would try to connect to the
Internet, or fail if the Internet wasn't available:

*   `check_updates_with_update`
*   `check_updates_none`
*   `check_updates_some`

This is because all three tests would run `rustup check`, and because the tests
weren't overriding the value of `RUSTUP_UPDATE_ROOT`, they'd go out to the
Internet to try to download self-updates.

The new helper mimics the behaviour of the `self_update_setup()` helper in the
same file.  It creates a mock self-update server in a local directory, and
points rustup to the directory using a `file://` URL.  When it tries to get
self-updates as part of `rustup check`, it only fetches from that directory.

kinnison reviewed (changes requested):
This is a really good start.  Thank you for wanting to tackle this issue.

I'm a little averse to seeing quite so much code duplication in the setup function though, could you look into refactoring the two functions into a common setup function and then two entry points (the existing function names) which call the common code with appropriate parameters?

alexwlchan commented:
> I'm a little averse to seeing quite so much code duplication in the setup function though, could you look into refactoring the two functions into a common setup function and then two entry points (the existing function names) which call the common code with appropriate parameters?

Definitely! I had a go at de-duplicating last night and couldn't get it working immediately, so I figured better to check I was on the right track before diving into the weeds on that one. I'll look at that later.

alexwlchan committed:
Consolidate the creation of the local self-update server

This reduces the repetition in the two "setup" methods.  Unfortunately, we
can't move the creation of the temp directory into the helper:

*   If it's created in the helper and only `self_dist` is returned, then
    `self_dist_tmp` will be dropped and the directory will be deleted.
*   If it's created in the helper and we use `into_path()` to persist it,
    then it won't be cleaned up when the function exits.

kinnison reviewed (approved):
To my eyes this is looking really good.

Thank you for doing this.
